Support for automatic usage and log uploads.

psigen opened the pull request:
This PR adds automatic logging of server usage and log synchronization via `Firebase`.

1) All starts and stop of the server are logged to a local database which is synchronized automatically the next time that the application is run while on wifi.
2) Each server has an entry containing version information and a "last update" timestamp.
3) The settings panel now has options to manually and automatically upload logfiles to the cloud.

This PR should be merged after #28.

psigen committed:
Added initial settings panel fragment.

psigen committed:
Added initial implementation of one-touch launcher.

psigen committed:
Added debugging panel.

psigen committed:
Using only directory based gitignore.

psigen committed:
Refactored LauncherActivity to deconflict other launchers.

I will need to create a few different "Launcher" activities,
so this change makes it clearer which is which.

psigen committed:
Added correct preference listings.

This change adds the necessary settings for the vehicle server to the
application preferences.

psigen committed:
Created new top-level activity to start the vehicle service.

psigen committed:
Added necessary settings to launcher, but layout is still very rough.

psigen committed:
Added working home location setter.

psigen committed:
Created simple placeholder for launching system.

psigen committed:
Added notification icon with Platypus logo.

psigen committed:
Removed now unused control interface stuff.

Will need to reimplement failsafe service, but for now this gets rid
of a lot of weird and deprecated code that just breaks with the new
service changes.

psigen committed:
Temporary changes to get the AirboatService to run at all.

Most of these changes will get undone later when the PCB controller
code is repaired.

psigen committed:
WHY WAS THIS ONE FILE USING TABS INSTEAD OF SPACES???

psigen committed:
Added a fancy launcher button.

It doesn't quite work yet, but it's promising!

psigen committed:
Added right/left indicators and fixed back button and theme.

psigen committed:
Added failsafe timeout to preferences.

psigen committed:
Added placeholder text to debug panel.

psigen committed:
Colorized start/stop button.

psigen committed:
Updated some more files and refactored logging.

psigen committed:
Added a preference for the server port setting.

psigen committed:
Half finished changes to make vehicle server work.

Currently stuck on how to abstract the controller such that
it doesn't need a context reference from any particular service.

psigen committed:
Merge branch 'feature/onetouch' of github.com:platypusllc/server into feature/onetouch

psigen committed:
Added shared preference listener placeholder.

psigen committed:
Added complete implementation of Controller class with instance().

psigen committed:
Refactored to use controller instance.

psigen committed:
Refactored classes and fixed controller references.

This commit repairs the references to controller data transmission and numerous logging references in the VehicleServerImpl.

It is now sort of working, except that the controller reference is not opened yet.

psigen committed:
Added better open/close semantics to enable clean shutdown.

psigen committed:
Merge branch 'feature/onetouch' of github.com:platypusllc/server into feature/onetouch

psigen committed:
Created a teleop view that allows thrust + rudder control.

psigen committed:
Added skeleton to actually debug panel to server.

psigen committed:
Fixed manifest paths, names, and intent filters.

psigen committed:
Cleaned up controller reference in Service.

(It's actually connecting to the board now!)

psigen committed:
Fixed issues with disconnect crashing the server.

psigen committed:
Renamed `ApplicationGlobe` to `PlatypusApplication`.

psigen committed:
Restructured controller to non-singleton architecture.

psigen committed:
Changed log level of disconnection event.

psigen committed:
Fixed missing connection events issue.

psigen committed:
Renamed activity that detects controller connection events.

psigen committed:
Added intent broadcast for service startup.

psigen committed:
Adjusted colors of startup button.

psigen committed:
Changed text on home button.

psigen committed:
Fixed "set home"  button to do something useful.

psigen committed:
Fixed pose logging and sending.

* Added `JSONArray` wrappers to position and rotation.
* Added missing `sendState()` calls to pose updates.

psigen committed:
Merge pull request #10 from platypusllc/bugfix/pose

Fixed server pose logging and sending.

Zeshengx committed:
Update TeleopView.java

psigen committed:
Merge pull request #12 from platypusllc/bugfix/rudder

Update TeleopView.java to fix rudder reverse

psigen committed:
Adds IP address dialog box to onetouch server.

This PR is a refactor of work by @zeshengx in #13.
> Add IP address showing in the LauncherFragment. The IP address will be checked every time when starting the server.

psigen committed:
Merge pull request #14 from platypusllc/feature/ip_address

Adds IP address dialog box to onetouch server.

psigen committed:
Added circle-ci support.

This commit adds a `circle.yml` and updated `build.gradle` to
use the newest tooling to allow automatic builds via circle-ci.

The resulting server releases are available as build artifacts
on the Circle-CI project page.

psigen committed:
Better output artifacts location.

psigen committed:
Merge pull request #16 from platypusllc/feature/circleci

Added circle-ci support.

psigen committed:
Fixed case where no location provider is found.

Previously if `getBestProvider()` returned `null`, indicating that no provider had sufficient accuracy, the server would crash.  This instead sends a warning to the user and shuts down the server.

psigen committed:
Fixed whitespace and cleaned up branching.

psigen committed:
Merge pull request #17 from platypusllc/bugfix/no_provider

Fixed case where no location provider is found.

christomaszewski committed:
Fixes to GPS permissions and default PID gains. (#21)

* Added permissions request for Location permissions in main activity.
* Added try-catch blocks to catch security exceptions in the vehicle service.
* Changed safe vehicle thrust to full speed.
* Updated default vehicle gains to new values.
* Fixed bug in serialization of gains to log files.

psigen committed:
Updated `platypus-core.jar` to include retransmission timeout fix. (#22)

* Updated core JAR file to include retransmission timeout fix.
* Updated remaining support and annotation libraries.

psigen committed:
Add PID gains to the debug panel. (#15)

* Added a GUI for editing PID gains.
* Added refresh and save buttons to the gain GUI.
* Changed gradlew to be executable.
* Increased update rate of GUI teleop panel to 20Hz.

psigen committed:
Added a velocity shutdown failsafe. (#23)

This commit adds a fixed timeout to velocity commands that causes
the vehicle to stop moving if no command is specified within a given
time window.

This mirrors the failsafe velocity shutdown that is provided on the
Arduino controller, which shuts down if the board is disconnected.

psigen committed:
Updated `platypus-core.jar` to the latest `master`.

psigen committed:
Changed logging to directly print to file instead of using microlog. (#25)

Since we already route all vehicle logging through the VehicleLogger object,
using microlog does not really help us in any way.  This PR removes the
microlog4j dependency, and instead directly logs to a logfile directory.

This also changes logging to use a proper android directory request, so
logs will now be put into:
`/sdcard/Documents/platypus`

It also slightly changes formatting to put the full date at the top of
each logfile, instead of printing it on every line.  The timestamps used
in the log are relative to the start of the log, which makes the logs
much shorter.

christomaszewski committed:
bugfix\permissions (#27)

* Updated permissions request in main application.
* Changed Waypoint achievement threshold radius to 3m.
* Added try-catch statements to handle corrupted sensor data.

psigen committed:
Regenerated project using Android Studio 2.0.

psigen committed:
Added git VCS configuration.

psigen committed:
Removed ignorable library xml files.

psigen committed:
Changed project name to most common name based on GitHub checkout.

psigen committed:
Added firebase support to build process.

psigen committed:
Added firebase libraries to app build.gradle.

psigen committed:
Added basic usage tracking code to server.

psigen committed:
Added error message for firebase initialization.

psigen committed:
Reorganized usage events to use the same entry.

psigen committed:
Fixed a number of minor errors in usage log formatting.

psigen committed:
Cleaned up timestamping code.

psigen committed:
Cleaned up cruft in main application.

psigen committed:
Added log file uploading service.

psigen committed:
Fixed typo in build path.

psigen committed:
Cleaned up vehicle logger to expose log filename.

psigen committed:
Merge branch 'bugfix/buildfiles' of https://github.com/platypusllc/server into feature/firebase

psigen committed:
Refactored automatic log uploads to use JobScheduler.

psigen committed:
Minor cleanup to log class.

psigen committed:
Improvements to logfile uploading.

psigen committed:
Set correct repeat timing on persistent job.

psigen committed:
Fixed typo in the log file format string.

psigen committed:
Added better error reporting for the usage logging.

psigen committed:
Added minor changest that might improve platform compatibility.

psigen committed:
Added authentication to Firebase.
- Fixed MD5 hash check on log uploads.
- Added crash reporting support.
- Created utility class for Firebase operations.

psigen committed:
Merge branch 'feature/firebase' of https://github.com/platypusllc/server into feature/firebase

# Conflicts:
#	app/src/main/java/com/platypus/android/server/VehicleService.java

psigen committed:
Fixed a few syntax errors from the merge.

psigen committed:
Fixed syntax errors from bad merge.

psigen committed:
Changed UDP server to use port setting in preferences.

psigen committed:
Added swipe on/off instead of long press button.

psigen committed:
Trying to update to android 24.

psigen committed:
Switched over to better base project.

christomaszewski commented:
I think this can probably be merged after the bugfix is pulled into it since we basically ship versions of this anyway.

psigen committed:
Fix waypoint and firebase server crashes. (#33)

* Bugfix for getWaypoints method that causes phone crash when sending waypoints with first point within completion range of boat.
* Bugfix in FirebaseUtils that causes crash on startup due to null reference for failure handler. Shroud direction fix.
* Bugfix for startWaypoints method that potentially causes phone crash if controller is set to "PRIMITIVES".

psigen committed:
Added updated gains for vectored thrust. (#35)

These updated gains were used by @christomaszewski on the latest
generate of rudder vehicles, so we will use them as the new
default values.

psigen committed:
Added fix that correctly parses errors from Controller. (#37)

* Added fix that correctly parses errors from Controller.
* Also minor auto-formatting fixes on changed files.
* Constructs controller messages as single write. (This is a stopgap to prevent the likelihood of message fragmentation on a USB write, but it is not guaranteed!)

psigen committed:
Added firebase configuration to .gitignore.

psigen committed:
Updated database reference and fixed a crash if there were no logs.

psigen committed:
Removed spam from server when disconnected from hardware.

psigen committed:
Added display of current boat type to launcher (#38)

* Refactored GUI classes into `.gui` subpackage.
* Added visualization of boat type to server.
* Added source files used to generate boat graphics.

psigen committed:
Added better exception handling to log uploading.

psigen committed:
Added support for QR code authentication credentials. (#39)

This modifies the authentication code to allow users to scan QR codes
to authenticate to the Platypus Cloud.

christomaszewski commented:
This should be good to merge since we are using it across all customer boats now.
